### PR TITLE
feat: transaction metric

### PR DESCRIPTION
**Pull request by fatelei**
ref: https://github.com/cectc/dbpack/issues/<issueID>

### Ⅰ. Describe what this PR did

#54

### CODE REVIEW

File: pkg/dt/distributed_transaction_manger.go [hunk 1]
```diff
@@ -24,7 +24,11 @@ import (
 	"net/url"
 	"time"
 
+	"github.com/cectc/dbpack/pkg/filter/metrics"
+	"github.com/cectc/dbpack/pkg/proto"
+
 	"github.com/go-resty/resty/v2"
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: please format the import block

File: pkg/dt/distributed_transaction_manger.go [hunk 2]
```diff
@@ -378,6 +414,17 @@ func (manager *DistributedTransactionManager) processNextBranchSession(ctx conte
 			}
 		}
 	}
+
+	if status == api.Complete {
+		// for go 1.16 time.Now().UnixMilli is undefined
+		// we can use time.Now().UnixMilli for go 1.18
+		now := time.Now().Second() * 1000
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: use `time.CurrentTimeMillis()`

File: pkg/dt/distributed_transaction_manger.go [hunk 3]
```diff
@@ -110,16 +124,18 @@ func (manager *DistributedTransactionManager) Begin(ctx context.Context, transac
 	if err := manager.storageDriver.AddGlobalSession(ctx, gt); err != nil {
 		return "", err
 	}
+	proto.WithVariable(ctx, globalTransactionStartTime, time.Now())
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: GlobalSession beginTime has recorded start time.

File: pkg/dt/distributed_transaction_manger.go [hunk 4]
```diff
@@ -110,16 +124,18 @@ func (manager *DistributedTransactionManager) Begin(ctx context.Context, transac
...
+	metrics.GlobalTransactionCounter.WithLabelValues(manager.applicationID, transactionName, metrics.TransactionStatusActive).Inc()
 	manager.globalSessionQueue.AddAfter(gt, time.Duration(timeout)*time.Millisecond)
 	log.Infof("successfully begin global transaction xid = {%s}", gt.XID)
 	return xid, nil
 }
 
-func (manager *DistributedTransactionManager) Commit(ctx context.Context, xid string) (api.GlobalSession_GlobalStatus, error) {
+func (manager *DistributedTransactionManager) Commit(ctx context.Context, xid string, transactionName string) (api.GlobalSession_GlobalStatus, error) {
 	return manager.storageDriver.GlobalCommit(ctx, xid)
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: transactionName is unused.

File: pkg/dt/distributed_transaction_manger.go [hunk 5]
```diff
@@ -110,16 +124,18 @@ func (manager *DistributedTransactionManager) Begin(ctx context.Context, transac
...
 }
 
-func (manager *DistributedTransactionManager) Rollback(ctx context.Context, xid string) (api.GlobalSession_GlobalStatus, error) {
+func (manager *DistributedTransactionManager) Rollback(ctx context.Context, xid string, transactionName string) (api.GlobalSession_GlobalStatus, error) {
 	return manager.storageDriver.GlobalRollback(ctx, xid)
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: transactionName is unused.

File: pkg/dt/distributed_transaction_manger.go [hunk 6]
```diff
@@ -144,6 +160,7 @@ func (manager *DistributedTransactionManager) BranchRegister(ctx context.Context
 	if err := manager.storageDriver.AddBranchSession(ctx, bs); err != nil {
 		return "", 0, err
 	}
+	metrics.BranchTransactionCounter.WithLabelValues(manager.applicationID, in.ResourceID, metrics.TransactionStatusActive).Inc()
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go:150]: the status is no correct

File: pkg/dt/distributed_transaction_manger.go [hunk 7]
```diff
@@ -220,14 +237,15 @@ func (manager *DistributedTransactionManager) _branchRollback(bs *api.BranchSess
...
+	ctx := context.Background()
+	globalSessions, err := manager.storageDriver.ListGlobalSession(ctx, manager.applicationID)
 	if err != nil {
 		return err
 	}
 	for _, gs := range globalSessions {
 		if gs.Status == api.Begin {
 			if isGlobalSessionTimeout(gs) {
-				if _, err := manager.Rollback(context.Background(), gs.XID); err != nil {
+				if _, err := manager.Rollback(ctx, gs.XID, gs.TransactionName); err != nil {
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: transactionName is unused.

File: pkg/dt/distributed_transaction_manger.go [hunk 8]
```diff
@@ -277,7 +302,7 @@ func (manager *DistributedTransactionManager) processNextGlobalSession(ctx conte
 	}
 	if newGlobalSession.Status == api.Begin {
 		if isGlobalSessionTimeout(newGlobalSession) {
-			_, err := manager.Rollback(context.Background(), newGlobalSession.XID)
+			_, err = manager.Rollback(context.Background(), newGlobalSession.XID, gs.TransactionName)
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: transactionName is unused.

File: pkg/dt/distributed_transaction_manger.go [hunk 9]
```diff
@@ -350,8 +381,12 @@ func (manager *DistributedTransactionManager) processNextBranchSession(ctx conte
 	defer manager.branchSessionQueue.Done(obj)
 
 	bs := obj.(*api.BranchSession)
+	var status api.BranchSession_BranchStatus
+	var err error
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]:
```suggestion
	var (
	        status api.BranchSession_BranchStatus
	        transactionStatus string
	        err error
	)
```

File: pkg/dt/distributed_transaction_manger.go (hunk 6 above)
Comment (fatelei) [reply pkg/dt/distributed_transaction_manger.go:150]: i think when register a branch session, the branch transaction will be start

File: pkg/dt/distributed_transaction_manger.go [hunk 10]
```diff
@@ -238,12 +250,19 @@ func (manager *DistributedTransactionManager) processGlobalSessions() error {
...
 				if err := manager.storageDriver.DeleteGlobalSession(context.Background(), gs.XID); err != nil {
 					return err
 				}
 				log.Debugf("global session finished, key: %s", gs.XID)
 			}
+			switch gs.Status {
+			case api.Committing:
+				manager.recordGlobalTransactionMetric(gs.TransactionName, metrics.TransactionStatusCommitted)
+			case api.Rollbacking:
+				manager.recordGlobalTransactionMetric(gs.TransactionName, metrics.TransactionStatusRollback)
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]:
```suggestion
				manager.recordGlobalTransactionMetric(gs.TransactionName, metrics.TransactionStatusRollbacked)
```

File: pkg/dt/distributed_transaction_manger.go [hunk 11]
```diff
@@ -361,14 +392,15 @@ func (manager *DistributedTransactionManager) processNextBranchSession(ctx conte
 		}
 	}
 	if bs.Status == api.PhaseTwoRollbacking {
+		transactionStatus = metrics.TransactionStatusRollback
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]:
```suggestion
		transactionStatus = metrics.TransactionStatusRollbacked
```

File: pkg/filter/dt/transaction.go [hunk 12]
```diff
@@ -40,13 +40,15 @@ func (f *_httpFilter) handleHttp1GlobalBegin(ctx *fasthttp.RequestCtx, transacti
 			transactionInfo, err)
 	}
 	ctx.SetUserValue(XID, xid)
+	ctx.SetUserValue(TransactionName, transactionInfo.RequestPath)
```
Comment (dk-lockdown) [inline pkg/filter/dt/transaction.go]: transactionName is unused

File: pkg/filter/metrics/transaction_metric.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package metrics
+
+import "github.com/prometheus/client_golang/prometheus"
+
+const (
+	TransactionStatusActive    = "active"
+	TransactionStatusCommitted = "committed"
+	TransactionStatusRollback  = "rollback"
```
Comment (dk-lockdown) [inline pkg/filter/metrics/transaction_metric.go]:
```suggestion
	TransactionStatusRollbacked  = "rollbacked"
```

File: pkg/dt/distributed_transaction_manger.go [hunk 14]
```diff
@@ -90,11 +93,17 @@ type DistributedTransactionManager struct {
...
 
 	globalSessionQueue workqueue.DelayingInterface
 	branchSessionQueue workqueue.Interface
 }
 
+func (manager *DistributedTransactionManager) recordGlobalTransactionMetric(transactionName string, transactionStatus string) {
+	metrics.GlobalTransactionCounter.WithLabelValues(manager.applicationID, transactionName, metrics.TransactionStatusActive).Desc()
+	metrics.GlobalTransactionCounter.WithLabelValues(manager.applicationID, transactionName, transactionStatus).Inc()
+}
+
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: suggest move this func after func `watchBranchSession`

File: pkg/dt/distributed_transaction_manger.go [hunk 15]
```diff
@@ -238,12 +245,19 @@ func (manager *DistributedTransactionManager) processGlobalSessions() error {
...
 				return err
 			}
+			// branch session has been committed or rollbacked
 			if len(bsKeys) == 0 {
 				if err := manager.storageDriver.DeleteGlobalSession(context.Background(), gs.XID); err != nil {
 					return err
 				}
 				log.Debugf("global session finished, key: %s", gs.XID)
 			}
+			switch gs.Status {
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go]: this block should move in  `if len(bsKeys) == 0` block

File: pkg/filter/metrics/transaction_metric.go [hunk 16]
```diff
@@ -0,0 +1,38 @@
+package metrics
```
Comment (dk-lockdown) [inline pkg/filter/metrics/transaction_metric.go]: this file should be move to 'pkg/dt/metrics', please add license header

**Review by dk-lockdown — APPROVED** [review]

File: pkg/dt/distributed_transaction_manger.go [hunk 17]
```diff
@@ -238,11 +243,18 @@ func (manager *DistributedTransactionManager) processGlobalSessions() error {
...
 			}
+			// branch session has been committed or rollbacked
 			if len(bsKeys) == 0 {
 				if err := manager.storageDriver.DeleteGlobalSession(context.Background(), gs.XID); err != nil {
 					return err
 				}
 				log.Debugf("global session finished, key: %s", gs.XID)
+				switch gs.Status {
+				case api.Committing:
+					manager.recordGlobalTransactionMetric(gs.TransactionName, metrics.TransactionStatusCommitted)
```
Comment (zhu733756) [inline pkg/dt/distributed_transaction_manger.go:260]: if a transaction is timeouted, how do we count the metric?

File: pkg/dt/distributed_transaction_manger.go [hunk 18]
```diff
@@ -361,14 +385,15 @@ func (manager *DistributedTransactionManager) processNextBranchSession(ctx conte
 		}
 	}
 	if bs.Status == api.PhaseTwoRollbacking {
+		transactionStatus = metrics.TransactionStatusRollbacked
 		if manager.IsRollingBackDead(bs) {
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go:402]: need a metrics record rollback dead branch session

File: pkg/dt/distributed_transaction_manger.go [hunk 19]
```diff
@@ -220,14 +224,15 @@ func (manager *DistributedTransactionManager) _branchRollback(bs *api.BranchSess
...
 func (manager *DistributedTransactionManager) processGlobalSessions() error {
-	globalSessions, err := manager.storageDriver.ListGlobalSession(context.Background(), manager.applicationID)
+	ctx := context.Background()
+	globalSessions, err := manager.storageDriver.ListGlobalSession(ctx, manager.applicationID)
 	if err != nil {
 		return err
 	}
 	for _, gs := range globalSessions {
 		if gs.Status == api.Begin {
 			if isGlobalSessionTimeout(gs) {
```
Comment (dk-lockdown) [inline pkg/dt/distributed_transaction_manger.go:234]: need a metrics record global session timeout

File: pkg/dt/distributed_transaction_manger.go (hunk 17 above)
Comment (dk-lockdown) [reply pkg/dt/distributed_transaction_manger.go:260]: good question

**Comment by zhu733756** [comment]
LGTM

**Review by hzliangbin — APPROVED** [review]

**Comment by hzliangbin** [comment]
LGTM

**Pull request merged**